Spring Data:  `quarkus.hibernate-orm.sql-load-script` is run twice after 47338 PR got merged

### Describe the bug

I use `quarkus.hibernate-orm.sql-load-script=import.sql` in my Spring Data application and it seems that when https://github.com/quarkusio/quarkus/pull/47338 got merged, the script is run more than once, otherwise I don't understand the error messages.

### Expected behavior

Works as before.

### Actual behavior

I can see logged: 

```
19:59:19,688 INFO  [app] 19:59:18,915 GenerationTarget encountered exception accepting command : Error executing DDL "INSERT INTO address(id, zipCode) VALUES (1, '28080')" via JDBC [ERROR: duplicate key value violates unique constraint "address_pkey"
19:59:19,688 INFO  [app]   Detail: Key (id)=(1) already exists.]: org.hibernate.tool.schema.spi.CommandAcceptanceException: Error executing DDL "INSERT INTO address(id, zipCode) VALUES (1, '28080')" via JDBC [ERROR: duplicate key value violates unique constraint "address_pkey"
19:59:19,688 INFO  [app]   Detail: Key (id)=(1) already exists.]
19:59:19,688 INFO  [app] 	at org.hibernate.tool.schema.internal.exec.GenerationTargetToDatabase.accept(GenerationTargetToDatabase.java:93)
19:59:19,689 INFO  [app] 	at org.hibernate.tool.schema.internal.Helper.applySqlString(Helper.java:221)
19:59:19,689 INFO  [app] 	at org.hibernate.tool.schema.internal.Helper.applyScript(Helper.java:244)
19:59:19,689 INFO  [app] 	at org.hibernate.tool.schema.internal.AbstractSchemaPopulator.applyImportFiles(AbstractSchemaPopulator.java:150)
19:59:19,689 INFO  [app] 	at org.hibernate.tool.schema.internal.AbstractSchemaPopulator.applyImportSources(AbstractSchemaPopulator.java:58)
19:59:19,689 INFO  [app] 	at org.hibernate.tool.schema.internal.SchemaCreatorImpl.performCreation(SchemaCreatorImpl.java:172)
19:59:19,689 INFO  [app] 	at org.hibernate.tool.schema.internal.SchemaCreatorImpl.doCreation(SchemaCreatorImpl.java:130)
19:59:19,690 INFO  [app] 	at org.hibernate.tool.schema.internal.SchemaCreatorImpl.doCreation(SchemaCreatorImpl.java:106)
19:59:19,690 INFO  [app] 	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.performDatabaseAction(SchemaManagementToolCoordinator.java:237)
19:59:19,690 INFO  [app] 	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.lambda$process$5(SchemaManagementToolCoordinator.java:142)
19:59:19,690 INFO  [app] 	at java.base/java.util.HashMap.forEach(HashMap.java:1429)
19:59:19,690 INFO  [app] 	at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.process(SchemaManagementToolCoordinator.java:139)
19:59:19,691 INFO  [app] 	at io.quarkus.hibernate.orm.runtime.observers.SessionFactoryObserverForSchemaExport.sessionFactoryCreated(SessionFactoryObserverForSchemaExport.java:21)
19:59:19,691 INFO  [app] 	at org.hibernate.internal.SessionFactoryObserverChain.sessionFactoryCreated(SessionFactoryObserverChain.java:33)
19:59:19,691 INFO  [app] 	at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:350)
19:59:19,691 INFO  [app] 	at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:92)
19:59:19,691 INFO  [app] 	at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:74)
19:59:19,692 INFO  [app] 	at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:90)
19:59:19,692 INFO  [app] 	at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:66)
19:59:19,693 INFO  [app] 	at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:165)
19:59:19,694 INFO  [app] 	at io.quarkus.hibernate.orm.runtime.JPAConfig$1.run(JPAConfig.java:61)
19:59:19,694 INFO  [app] 	at java.base/java.lang.Thread.run(Thread.java:1583)
```


### How to Reproduce?

Reproducer:

1. `git clone git@github.com:quarkus-qe/quarkus-test-suite.git`
2. `cd quarkus-test-suite/spring/spring-data`
3. `mvn clean verify -Dit.test=PagingAndSortingRepositoryRestResourceIT#create`


### Output of `uname -a` or `ver`

Fedora 41, Ubuntu latest

### Output of `java -version`

Temurin 21 and 17

### Quarkus version or git rev

999-SNAPSHOT

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.9.9

### Additional information

_No response_

I am not sure if this is an issue with the Hibernate extension, or how it is actually being used in the TestSuite.
What I do know for sure is that it has nothing to do with Spring Data.

I have opened https://github.com/quarkusio/quarkus/pull/49087 to help the Hibernate folks debug this. That test doesn't fail, but you can see:

```
violation: "PRIMARY KEY ON PUBLIC.SHAREDENTITY(ID) ( /* key:1 */ CAST(1 AS BIGINT), 'default sql load script entity')";]: org.hibernate.tool.schema.spi.CommandAcceptanceException: Error executing DDL "INSERT INTO SharedEntity(id, name) VALUES(1, 'default sql load script entity')" via JDBC [Unique index or primary key violation: "PRIMARY KEY ON PUBLIC.SHAREDENTITY(ID) ( /* key:1 */ CAST(1 AS BIGINT), 'default sql load script entity')";]
        at org.hibernate.tool.schema.internal.exec.GenerationTargetToDatabase.accept(GenerationTargetToDatabase.java:93)
        at org.hibernate.tool.schema.internal.Helper.applySqlString(Helper.java:221)
        at org.hibernate.tool.schema.internal.Helper.applyScript(Helper.java:244)
        at org.hibernate.tool.schema.internal.AbstractSchemaPopulator.applyImportFiles(AbstractSchemaPopulator.java:150)
        at org.hibernate.tool.schema.internal.AbstractSchemaPopulator.applyImportSources(AbstractSchemaPopulator.java:58)
        at org.hibernate.tool.schema.internal.SchemaCreatorImpl.performCreation(SchemaCreatorImpl.java:172)
        at org.hibernate.tool.schema.internal.SchemaCreatorImpl.doCreation(SchemaCreatorImpl.java:130)
        at org.hibernate.tool.schema.internal.SchemaCreatorImpl.doCreation(SchemaCreatorImpl.java:106)
        at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.performDatabaseAction(SchemaManagementToolCoordinator.java:237)
        at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.lambda$process$5(SchemaManagementToolCoordinator.java:142)
        at java.base/java.util.HashMap.forEach(HashMap.java:1429)
        at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.process(SchemaManagementToolCoordinator.java:139)
        at io.quarkus.hibernate.orm.runtime.observers.SessionFactoryObserverForSchemaExport.sessionFactoryCreated(SessionFactoryObserverForSchemaExport.java:21)
        at org.hibernate.internal.SessionFactoryObserverChain.sessionFactoryCreated(SessionFactoryObserverChain.java:33)
        at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:350)
        at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:92)
        at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:74)
        at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:90)
        at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:66)
        at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:165)
        at io.quarkus.hibernate.orm.runtime.JPAConfig$1.run(JPAConfig.java:61)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.h2.jdbc.JdbcSQLIntegrityConstraintViolationException: Unique index or primary key violation: "PRIMARY KEY ON PUBLIC.SHAREDENTITY(ID) ( /* key:1 */ CAST(1 AS BIGINT), 'default sql load script entity')"; SQL statement:
INSERT INTO SharedEntity(id, name) VALUES(1, 'default sql load script entity') [23505-230]
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:520)
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:489)
        at org.h2.message.DbException.get(DbException.java:223)
        at org.h2.message.DbException.get(DbException.java:199)
        at org.h2.mvstore.db.MVPrimaryIndex.add(MVPrimaryIndex.java:121)
        at org.h2.mvstore.db.MVTable.addRow(MVTable.java:518)
        at org.h2.command.dml.Insert.insertRows(Insert.java:174)
        at org.h2.command.dml.Insert.update(Insert.java:135)
        at org.h2.command.dml.DataChangeStatement.update(DataChangeStatement.java:74)
        at org.h2.command.CommandContainer.update(CommandContainer.java:139)
        at org.h2.command.Command.executeUpdate(Command.java:304)
        at org.h2.command.Command.executeUpdate(Command.java:248)
        at org.h2.server.TcpServerThread.process(TcpServerThread.java:386)
        at org.h2.server.TcpServerThread.run(TcpServerThread.java:193)
        at java.base/java.lang.Thread.run(Thread.java:1583)

        at org.h2.message.DbException.getJdbcSQLException(DbException.java:520)
        at org.h2.engine.SessionRemote.readSQLException(SessionRemote.java:663)
        at org.h2.engine.SessionRemote.readException(SessionRemote.java:646)
        at org.h2.engine.SessionRemote.done(SessionRemote.java:620)
        at org.h2.command.CommandRemote.executeUpdate(CommandRemote.java:225)
        at org.h2.jdbc.JdbcStatement.executeInternal(JdbcStatement.java:262)
        at org.h2.jdbc.JdbcStatement.execute(JdbcStatement.java:231)
        at io.agroal.pool.wrapper.StatementWrapper.execute(StatementWrapper.java:220)
        at org.hibernate.tool.schema.internal.exec.GenerationTargetToDatabase.accept(GenerationTargetToDatabase.java:79)
        ... 21 more

2025-07-24 11:52:41,644 WARN  [org.hib.too.sch.int.ExceptionHandlerLoggedImpl] (JPA Startup Thread: users) GenerationTarget encountered exception accepting command : Error executing DDL "INSERT INTO SharedEntity(id, name) VALUES(2, 'import.sql load script entity')" via JDBC [Unique index or primary key violation: "PRIMARY KEY ON PUBLIC.SHAREDENTITY(ID) ( /* key:2 */ CAST(2 AS BIGINT), 'import.sql load script entity')";]: org.hibernate.tool.schema.spi.CommandAcceptanceException: Error executing DDL "INSERT INTO SharedEntity(id, name) VALUES(2, 'import.sql load script entity')" via JDBC [Unique index or primary key violation: "PRIMARY KEY ON PUBLIC.SHAREDENTITY(ID) ( /* key:2 */ CAST(2 AS BIGINT), 'import.sql load script entity')";]
        at org.hibernate.tool.schema.internal.exec.GenerationTargetToDatabase.accept(GenerationTargetToDatabase.java:93)
        at org.hibernate.tool.schema.internal.Helper.applySqlString(Helper.java:221)
        at org.hibernate.tool.schema.internal.Helper.applyScript(Helper.java:244)
        at org.hibernate.tool.schema.internal.AbstractSchemaPopulator.applyImportFiles(AbstractSchemaPopulator.java:150)
        at org.hibernate.tool.schema.internal.AbstractSchemaPopulator.applyImportSources(AbstractSchemaPopulator.java:58)
        at org.hibernate.tool.schema.internal.SchemaCreatorImpl.performCreation(SchemaCreatorImpl.java:172)
        at org.hibernate.tool.schema.internal.SchemaCreatorImpl.doCreation(SchemaCreatorImpl.java:130)
        at org.hibernate.tool.schema.internal.SchemaCreatorImpl.doCreation(SchemaCreatorImpl.java:106)
        at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.performDatabaseAction(SchemaManagementToolCoordinator.java:237)
        at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.lambda$process$5(SchemaManagementToolCoordinator.java:142)
        at java.base/java.util.HashMap.forEach(HashMap.java:1429)
        at org.hibernate.tool.schema.spi.SchemaManagementToolCoordinator.process(SchemaManagementToolCoordinator.java:139)
        at io.quarkus.hibernate.orm.runtime.observers.SessionFactoryObserverForSchemaExport.sessionFactoryCreated(SessionFactoryObserverForSchemaExport.java:21)
        at org.hibernate.internal.SessionFactoryObserverChain.sessionFactoryCreated(SessionFactoryObserverChain.java:33)
        at org.hibernate.internal.SessionFactoryImpl.<init>(SessionFactoryImpl.java:350)
        at io.quarkus.hibernate.orm.runtime.boot.FastBootEntityManagerFactoryBuilder.build(FastBootEntityManagerFactoryBuilder.java:92)
        at io.quarkus.hibernate.orm.runtime.FastBootHibernatePersistenceProvider.createEntityManagerFactory(FastBootHibernatePersistenceProvider.java:74)
        at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:90)
        at jakarta.persistence.Persistence.createEntityManagerFactory(Persistence.java:66)
        at io.quarkus.hibernate.orm.runtime.JPAConfig$LazyPersistenceUnit.get(JPAConfig.java:165)
        at io.quarkus.hibernate.orm.runtime.JPAConfig$1.run(JPAConfig.java:61)
        at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.h2.jdbc.JdbcSQLIntegrityConstraintViolationException: Unique index or primary key violation: "PRIMARY KEY ON PUBLIC.SHAREDENTITY(ID) ( /* key:2 */ CAST(2 AS BIGINT), 'import.sql load script entity')"; SQL statement:
INSERT INTO SharedEntity(id, name) VALUES(2, 'import.sql load script entity') [23505-230]
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:520)
        at org.h2.message.DbException.getJdbcSQLException(DbException.java:489)
        at org.h2.message.DbException.get(DbException.java:223)
        at org.h2.message.DbException.get(DbException.java:199)
        at org.h2.mvstore.db.MVPrimaryIndex.add(MVPrimaryIndex.java:121)
        at org.h2.mvstore.db.MVTable.addRow(MVTable.java:518)
        at org.h2.command.dml.Insert.insertRows(Insert.java:174)
        at org.h2.command.dml.Insert.update(Insert.java:135)
        at org.h2.command.dml.DataChangeStatement.update(DataChangeStatement.java:74)
        at org.h2.command.CommandContainer.update(CommandContainer.java:139)
        at org.h2.command.Command.executeUpdate(Command.java:304)
        at org.h2.command.Command.executeUpdate(Command.java:248)
        at org.h2.server.TcpServerThread.process(TcpServerThread.java:386)
        at org.h2.server.TcpServerThread.run(TcpServerThread.java:193)
        at java.base/java.lang.Thread.run(Thread.java:1583)

        at org.h2.message.DbException.getJdbcSQLException(DbException.java:520)
        at org.h2.engine.SessionRemote.readSQLException(SessionRemote.java:663)
        at org.h2.engine.SessionRemote.readException(SessionRemote.java:646)
        at org.h2.engine.SessionRemote.done(SessionRemote.java:620)
        at org.h2.command.CommandRemote.executeUpdate(CommandRemote.java:225)
        at org.h2.jdbc.JdbcStatement.executeInternal(JdbcStatement.java:262)
        at org.h2.jdbc.JdbcStatement.execute(JdbcStatement.java:231)
        at io.agroal.pool.wrapper.StatementWrapper.execute(StatementWrapper.java:220)
        at org.hibernate.tool.schema.internal.exec.GenerationTargetToDatabase.accept(GenerationTargetToDatabase.java:79)
        ... 21 more
```

which is essentially the same cause as the issue in the TestSuite

--- a/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/util/HibernateProcessorUtil.java
+++ b/extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/util/HibernateProcessorUtil.java
@@ -4,6 +4,7 @@ import static io.quarkus.hibernate.orm.deployment.HibernateConfigUtil.firstPrese
 
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Locale;
@@ -330,9 +331,10 @@ public final class HibernateProcessorUtil {
             BuildProducer<NativeImageResourceBuildItem> nativeImageResources,
             BuildProducer<HotDeploymentWatchedFileBuildItem> hotDeploymentWatchedFiles,
             QuarkusPersistenceUnitDescriptor descriptor) {
-        // sql-load-scripts
+        // This defaults to 'import.sql' in non-production modes
         List<String> importFiles = getSqlLoadScript(persistenceUnitConfig.sqlLoadScript(), launchMode);
         if (!importFiles.isEmpty()) {
+            List<String> existingImportFiles = new ArrayList<>();
             for (String importFile : importFiles) {
                 Path loadScriptPath;
                 try {
@@ -347,6 +349,7 @@ public final class HibernateProcessorUtil {
 
                 if (loadScriptPath != null && !Files.isDirectory(loadScriptPath)) {
                     // enlist resource if present
+                    existingImportFiles.add(importFile);
                     nativeImageResources.produce(new NativeImageResourceBuildItem(importFile));
                 } else if (persistenceUnitConfig.sqlLoadScript().isPresent()) {
                     //raise exception if explicit file is not present (i.e. not the default)
@@ -361,15 +364,13 @@ public final class HibernateProcessorUtil {
                 hotDeploymentWatchedFiles.produce(new HotDeploymentWatchedFileBuildItem(importFile));
             }
 
-            // only set the found import files if configured
-            if (persistenceUnitConfig.sqlLoadScript().isPresent()) {
+            if (!existingImportFiles.isEmpty()) {
                 descriptor.getProperties().setProperty(AvailableSettings.JAKARTA_HBM2DDL_LOAD_SCRIPT_SOURCE,
-                        String.join(",", importFiles));
+                        String.join(",", existingImportFiles));
             }
-        } else {
-            //Disable implicit loading of the default import script (import.sql)
-            descriptor.getProperties().setProperty(AvailableSettings.JAKARTA_HBM2DDL_LOAD_SCRIPT_SOURCE, "");
-            descriptor.getProperties().setProperty(AvailableSettings.HBM2DDL_SKIP_DEFAULT_IMPORT_FILE, "true");
         }
+
+        //Disable implicit loading of the default import script (import.sql)
+        descriptor.getProperties().setProperty(AvailableSettings.HBM2DDL_SKIP_DEFAULT_IMPORT_FILE, "true");
     }
 }

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/sql_load_script/DefaultSqlLoadScriptAbsentTestCase.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/sql_load_script/DefaultSqlLoadScriptAbsentTestCase.java
@@ -1,0 +1,30 @@
+package io.quarkus.hibernate.orm.sql_load_script;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.logging.Level;
+import java.util.logging.LogRecord;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.hibernate.orm.MyEntity;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class DefaultSqlLoadScriptAbsentTestCase {
+    @RegisterExtension
+    static QuarkusUnitTest runner = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(MyEntity.class, SqlLoadScriptTestResource.class)
+                    .addAsResource("application.properties"))
+            .setLogRecordPredicate(record -> record.getLevel().intValue() >= Level.WARNING.intValue())
+            // In particular, we don't want Hibernate ORM to log
+            // "Specified schema generation script file [import.sql] did not exist for reading"
+            // when "import.sql" is just the Quarkus default.
+            .assertLogRecords(records -> assertThat(records).extracting(LogRecord::getMessage).isEmpty());
+
+    @Test
+    public void testImportSqlLoadScriptTest() {
+        // No startup failure, so we're already good.
+    }
+}

--- a/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/sql_load_script/DefaultSqlLoadScriptUnusedTestCase.java
+++ b/extensions/hibernate-orm/deployment/src/test/java/io/quarkus/hibernate/orm/sql_load_script/DefaultSqlLoadScriptUnusedTestCase.java
@@ -1,0 +1,39 @@
+package io.quarkus.hibernate.orm.sql_load_script;
+
+import org.hamcrest.Matchers;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.hibernate.orm.MyEntity;
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.RestAssured;
+
+/**
+ * Test that import.sql is not executed when a different file is specified in quarkus.hibernate-orm.sql-load-file.
+ *
+ * This could happen if 'import.sql' is used for a different persistence unit, for example.
+ *
+ * See https://github.com/quarkusio/quarkus/issues/49075
+ */
+public class DefaultSqlLoadScriptUnusedTestCase {
+    @RegisterExtension
+    static QuarkusUnitTest runner = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(MyEntity.class, SqlLoadScriptTestResource.class)
+                    .addAsResource("application.properties")
+                    // If both these files get executed, a constraint failure will occur due to duplicate PKs.
+                    .addAsResource("import.sql")
+                    .addAsResource("import.sql", "import-different-name.sql"))
+            .overrideConfigKey("quarkus.hibernate-orm.sql-load-script", "import-different-name.sql")
+            // Fail startup if schema management has errors (like... executing the load script multiple times)
+            .overrideConfigKey("quarkus.hibernate-orm.schema-management.halt-on-error", "true");
+
+    @Test
+    public void testImportSqlLoadScriptTest() {
+        // No startup failure, so we're already good.
+
+        // Let's just check the script _was_ executed.
+        String name = "import.sql load script entity";
+        RestAssured.when().get("/orm-sql-load-script/2").then().body(Matchers.is(name));
+    }
+}

